### PR TITLE
spanner-jdbc: Step 24 - Parameter store for jdbc parameters

### PR DESCRIPTION
**Pull request by olavloite**
Adds a generic data store for JDBC parameters. JDBC parameters are position based, and can be set in any order. They are therefore stored in an ArrayList that may contain null values for parameters that have not yet been set.

### CODE REVIEW

File: google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/JdbcParameterStoreTest.java [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+    Statement.Builder builder = Statement.newBuilder("SELECT * FROM FOO WHERE BAR=:p1");
+    params.bindParameterValue(builder.bind("p1"), 1);
+    assertThat(builder.build().getParameters().get("p1"), is(equalTo(value)));
+  }
+
+  @Test
+  public void testConvertPositionalParametersToNamedParameters() throws SQLException {
+    assertThat(
+        convertPositionalParametersToNamedParameters(
+            "select * from foo where name=?").sqlWithNamedParameters,
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/JdbcParameterStoreTest.java]: can we add some more basic tests like this statement?

Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/JdbcParameterStoreTest.java]: Added additional test statements. Is there anything specific that you had in mind?

Comment (kolea2) [reply google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/JdbcParameterStoreTest.java]: That looks good, thanks!

File: google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/JdbcParameterStoreTest.java [hunk 2]
```diff
@@ -0,0 +1,512 @@
...
+    assertThat((JdbcArray) params.getParameter(1),
+        is(equalTo(JdbcArray.createArray("BOOL", new Boolean[] {true, false, true}))));
+    verifyParameter(params, Value.boolArray(new boolean[] {true, false, true}));
+
+    params.setParameter(1, JdbcArray.createArray("INT64", new Long[] {1L, 2L, 3L}), Types.ARRAY);
+    assertThat((JdbcArray) params.getParameter(1),
+        is(equalTo(JdbcArray.createArray("INT64", new Long[] {1L, 2L, 3L}))));
+    verifyParameter(params, Value.int64Array(new long[] {1, 2, 3}));
+
+    params.setParameter(1, JdbcArray.createArray("FLOAT64", new Double[] {1D, 2D, 3D}),
```
Comment (kolea2) [inline google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/JdbcParameterStoreTest.java]: can we add in some additional test cases with null values?

Comment (olavloite) [reply google-cloud-clients/google-cloud-contrib/google-cloud-spanner-jdbc/src/test/java/com/google/cloud/spanner/jdbc/JdbcParameterStoreTest.java]: Added tests with null values in the array, as well as setting a null value as an array parameter.
